client/verta/verta/_internal_utils/_histogram_utils.py

```diff
@@ -0,0 +1,209 @@
+# -*- coding: utf-8 -*-
+
+from ..external import six
+
+
+def calculate_histogram(data):
+    """
+    Calculates a histogram for `data`.
+
+    Parameters
+    ----------
+    data : pandas.Series
+        Data to be binned.
+
+    Returns
+    -------
+    histogram : dict
+
+    """
+    try:  # binary
+        return calculate_binary_histogram(data)
+    except HistogramError:
+        pass
+
+    try:  # discrete
+        return calculate_discrete_histogram(data)
+    except HistogramError:
+        pass
+
+    # continuous
+    return calculate_float_histogram(data)
+
+def calculate_binary_histogram(data):
+    """
+    Calculates a histogram for binary `data`.
+
+    Parameters
+    ----------
+    data : pandas.Series
+        Binary data to be binned.
+
+    Returns
+    -------
+    histogram : dict
+
+    Raises
+    ------
+    HistogramError
+        If a binary histogram cannot be calculated from `data`.
+
+    """
+    values = data.values.tolist()
+
+    zeros = 0
+    ones = 0
+    for value in values:
+        if isinstance(value, bool):
+            if value == False:
+                zeros += 1
+                continue
+            elif value == True:
+                ones += 1
+                continue
+
+        if isinstance(value, six.string_types):
+            # handle bool-like strings
+            if value.lower() == "false":
+                zeros += 1
+                continue
+            elif value.lower() == "true":
+                ones += 1
+                continue
+
+            # handle num-like strings (falls through to numeric case)
+            try:
+                value = float(value)
+            except ValueError:
+                pass
+
+        if isinstance(value, (six.integer_types, float)):
+            if value == 0:
+                zeros += 1
+                continue
+            elif value == 1:
+                ones += 1
+                continue
+
+        # unsupported value
+        raise HistogramError("invalid binary value {}".format(value))
+
+    return {
+        'histogram': {
+            'binary': {
+                'count': [zeros, ones],
+            },
+        },
+        'type': "binary",
+    }
+
+def calculate_discrete_histogram(data):
+    """
+    Calculates a histogram for discrete `data`.
+
+    Parameters
+    ----------
+    data : pandas.Series of int
+        Discrete data to be binned.
+
+    Returns
+    -------
+    histogram : dict
+
+    Raises
+    ------
+    HistogramError
+        If a discrete histogram cannot be calculated from `data`.
+
+    """
+    value_counts = data.value_counts().sort_index()
+    values = value_counts.index.tolist()
```

Comment:
I have no idea if the values will actually become the index here.

Reply:
That's what `pd.Series.value_counts()`'s return is: `.index` is the unique values from the original data, and `.values` are the counts of each unique value.

I can rename the variable `values` to `unique_values` if that would clarify this.

Reply:
It's fine, I just didn't know.